build_tools/cmake/iree_macros.cmake

```diff
@@ -106,45 +106,32 @@ endfunction()
 #
 # Gets the path to an executable in a cross-compilation-aware way. This
 # should be used when accessing binaries that are used as part of the build,
-# such as for generating files used for later build steps. Those binaries
-# can come from third-party projects or another CMake invocation.
+# such as for generating files used for later build steps.
 #
 # Paramters:
 # - OUTPUT_PATH_VAR: variable name for receiving the path to the built target.
 # - EXECUTABLE: the executable to get its path.
 function(iree_get_executable_path OUTPUT_PATH_VAR EXECUTABLE)
-  if(CMAKE_CROSSCOMPILING)
-    # The target is defined in the CMake invocation for host. We don't have
-    # access to the target; relying on the path here.
-    set(_OUTPUT_PATH "${IREE_HOST_BINARY_ROOT}/bin/${EXECUTABLE}${IREE_HOST_EXECUTABLE_SUFFIX}")
-    set(${OUTPUT_PATH_VAR} "${_OUTPUT_PATH}" PARENT_SCOPE)
-  else()
-    # The target is defined in this CMake invocation. We can query the location
-    # directly from CMake.
+  if (NOT CMAKE_CROSSCOMPILING OR TARGET "${EXECUTABLE}")
+    # We can either expect the target to be defined as part of this CMake
+    # invocation (if not cross compiling) or the target is defined already.
     set(${OUTPUT_PATH_VAR} "$<TARGET_FILE:${EXECUTABLE}>" PARENT_SCOPE)
+  else()
+    # The target won't be directly defined by this CMake invocation so check
+    # for an already built executable at IREE_HOST_BINARY_ROOT. If we find it,
+    # add it as an imported target so it gets picked up on later invocations.
+    set(_EXECUTABLE_PATH "${IREE_HOST_BINARY_ROOT}/bin/${EXECUTABLE}${IREE_HOST_EXECUTABLE_SUFFIX}")
+    if (EXISTS ${_EXECUTABLE_PATH})
+      add_executable("${EXECUTABLE}" IMPORTED GLOBAL)
```

Comment:
I wonder if it would be better to configure these targets with something like a `create_host_tool` function that creates the target by importing a prebuilt binary if it doesn't already exist. A bit tricky because the natural place to do that would be where the binary is defined and we're likely to disable that whole file with our enable options.

Basically it seems weird that we're creating a target when we ask to use the tool instead of where we create the tool.

Reply:
I was thinking along those lines as well. We could rename part of this to something like `iree_declare_executable`... but as you point out, the normal place to put that would be where the executable is defined, when we really only want to check for an executable where it is depended upon when cross compiling.